### PR TITLE
Hydropower modifications

### PR DESCRIPTION
**Pull request by MohammedBasheer**
The units in hydropower calculations have been revised.
New parameters and recorders have been added to accept variable turbine_elevation in terms of a parameter.

### CODE REVIEW

File: pywr/recorders/_hydropower.pyx [hunk 1]
```diff
@@ -23,11 +23,11 @@ cpdef double hydropower_calculation(double flow, double water_elevation, double
...
     flow_unit_conversion : double (default=1.0)
         A factor used to transform the units of flow to be compatible with the equation here. This
-        should convert flow to units of $m^3/day$
+        should convert flow to units of $m^3/second$
     energy_unit_conversion : double (default=1e-6)
-        A factor used to transform the units of power. Defaults to 1e-6 to return $MJ$/day. 
+        A factor used to transform the units of power. Defaults to 1e-6 to return $MW$. 
     density : double (default=1000)
-        Density of water in $kgm^{-3}$.
+        Density of water in $kg/m^{-3}$.
```
Comment (snorfalorpagus) [inline pywr/recorders/_hydropower.pyx]: Density is a measure of unit mass per unit volume, so this should either be `kg / m^3` or `kg m^{-3}` as it was.

File: pywr/parameters/_hydropower.pyx [hunk 2]
```diff
@@ -23,17 +23,17 @@ cpdef double inverse_hydropower_calculation(double power, double water_elevation
         An efficiency scaling factor for the power output of the turbine.
     flow_unit_conversion : double (default=1.0)
         A factor used to transform the units of flow to be compatible with the equation here. This
-        should convert flow to units of $m^3/day$
+        should convert flow to units of $m^3/second$
     energy_unit_conversion : double (default=1e-6)
-        A factor used to transform the units of power. Defaults to 1e-6 to assuming input of $MJ$/day.
```
Comment (snorfalorpagus) [inline pywr/parameters/_hydropower.pyx]: @jetuk Do you have an opinion on the units we should be using?

File: pywr/recorders/_hydropower.pyx [hunk 3]
```diff
@@ -268,4 +387,125 @@ cdef class TotalHydroEnergyRecorder(BaseConstantNodeRecorder):
...
+        def __get__(self):
+            return self._turbine_elevation_parameter
+        def __set__(self, parameter):
+            if self._turbine_elevation_parameter:
+                self.children.remove(self._turbine_elevation_parameter)
+            self.children.add(parameter)
+            self._turbine_elevation_parameter = parameter 
+
+    cpdef after(self):
+        """ Calculate the  """
```
Comment (snorfalorpagus) [inline pywr/recorders/_hydropower.pyx]: TODO

File: pywr/recorders/_hydropower.pyx [hunk 4]
```diff
@@ -268,4 +387,125 @@ cdef class TotalHydroEnergyRecorder(BaseConstantNodeRecorder):
...
+        else:
+            water_elevation_parameter = None
+
+        if "turbine_elevation_parameter" in data:
+            turbine_elevation_parameter = load_parameter(model, data.pop("turbine_elevation_parameter"))
+        else:
+            turbine_elevation_parameter = None
+
+        return cls(model, node, water_elevation_parameter=water_elevation_parameter,turbine_elevation_parameter=turbine_elevation_parameter, **data)
+TotalHydroEnergyRecorderWithVaribaleTailwater.register()
```
Comment (snorfalorpagus) [inline pywr/recorders/_hydropower.pyx]: Missing newline at end of file.

File: pywr/parameters/_hydropower.pxd [hunk 5]
```diff
@@ -14,3 +14,16 @@ cdef class HydropowerTargetParameter(Parameter):
...
+    cdef Parameter _water_elevation_parameter
+    cdef Parameter _target
+    cdef Parameter _max_flow
+    cdef Parameter _min_flow
+    cdef public double min_head
+    cdef Parameter _turbine_elevation_parameter
+    cdef public double flow_unit_conversion
+    cdef public double energy_unit_conversion
+    cdef public double density
+    cdef public double efficiency
```
Comment (snorfalorpagus) [inline pywr/parameters/_hydropower.pxd]: Missing newline at end of file.

File: pywr/recorders/_hydropower.pyx [hunk 6]
```diff
@@ -164,6 +164,125 @@ cdef class HydropowerRecorder(NumpyArrayNodeRecorder):
 HydropowerRecorder.register()
 
 
+cdef class HydropowerRecorderWithVaribaleTailwater(NumpyArrayNodeRecorder):
```
Comment (snorfalorpagus) [inline pywr/recorders/_hydropower.pyx]: Please can you add some unit tests for the new functionality provided by these components?

File: pywr/recorders/_hydropower.pxd [hunk 7]
```diff
@@ -18,4 +18,22 @@ cdef class TotalHydroEnergyRecorder(BaseConstantNodeRecorder):
     cdef public double flow_unit_conversion
     cdef public double energy_unit_conversion
     cdef public double density
+    cdef public double efficiency
+
+
+    cdef class HydropowerRecorderWithVaribaleTailwater(NumpyArrayNodeRecorder):
```
Comment (snorfalorpagus) [inline pywr/recorders/_hydropower.pxd]: The indentation here is what is causing the current failures in the CI.

File: pywr/recorders/_hydropower.pyx (hunk 1 above)
Comment (MohammedBasheer) [reply pywr/recorders/_hydropower.pyx]: Yes, I agree.
I will correct that

File: pywr/parameters/_hydropower.pyx (hunk 2 above)
Comment (MohammedBasheer) [reply pywr/parameters/_hydropower.pyx]: MegaWatt (MW) is the unit commonly used for power

Comment (jetuk) [reply pywr/parameters/_hydropower.pyx]: Yeah, MW probably makes a better default to MJ.

**Pull request closed without merging**